File: Plugin/src/main/java/dev/rosewood/roseskyblock/world/generator/LayeredChunkGenerator.java
package dev.rosewood.roseskyblock.world.generator;

import org.bukkit.block.Biome;
import org.bukkit.generator.ChunkGenerator;
import org.bukkit.generator.WorldInfo;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Random;

public class LayeredChunkGenerator extends ChunkGenerator {

    private final List<ChunkLayer> layers;

    public LayeredChunkGenerator(List<ChunkLayer> layers) {
        this.layers = layers;
    }

    @Override
    public void generateSurface(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ, @NotNull ChunkGenerator.ChunkData chunkData) {
        this.layers.forEach(chunkLayer -> chunkLayer.fill(chunkData));
    }

    @Override
    public void generateBedrock(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ, @NotNull ChunkGenerator.ChunkData chunkData) {
        this.layers.forEach(chunkLayer -> chunkLayer.fill(chunkData));
    }

    @Override
    public void generateNoise(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ, @NotNull ChunkGenerator.ChunkData chunkData) {
        this.layers.forEach(chunkLayer -> chunkLayer.fill(chunkData));
    }

    @Override
    public void generateCaves(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ, @NotNull ChunkGenerator.ChunkData chunkData) {
        this.layers.forEach(chunkLayer -> chunkLayer.fill(chunkData));
    }

}

File: src/main/kotlin/dev/rosewood/roseskyblock/world/generator/LayeredChunkGenerator.kt
package dev.rosewood.roseskyblock.world.generator

import dev.rosewood.rosegarden.utils.NMSUtil
import java.util.Random
import org.bukkit.World
import org.bukkit.block.Biome
import org.bukkit.generator.ChunkGenerator

class LayeredChunkGenerator(private val chunkLayers: List<ChunkLayer>, private val biome: Biome) : ChunkGenerator() {

    override fun generateChunkData(world: World, random: Random, chunkX: Int, chunkZ: Int, biomeGrid: BiomeGrid): ChunkData {
        val chunkData = this.createChunkData(world)
        this.chunkLayers.forEach { it.fill(chunkData) }

        // 1.16 has 3d biomes, while older versions do not
        // 3d biomes are separated into 4x4 pieces, so there are 64 pieces per chunk section, or 1024 per chunk
        if (NMSUtil.getVersionNumber() >= 16) {
            for (x in 0..15 step 4)
                for (z in 0..15 step 4)
                    for (y in 0..world.maxHeight step 4)
                        biomeGrid.setBiome(x, y, z, this.biome)
        } else {
            @Suppress("DEPRECATION")
            for (x in 0..15)
                for (z in 0..15)
                    biomeGrid.setBiome(x, z, this.biome)
        }

        return chunkData
    }

    override fun isParallelCapable(): Boolean = true
}
